File: app/javascript/channels/search_channel.js
console.log("Hello from app/javascript/channels/search_channel.js");

import consumer from "channels/consumer";

const MINIMUM_QUERY_LENGTH = 3;
const BUFFERING_DELAY_MS = 300; // Adjust as needed

let bufferTimer;
const searchChannel = consumer.subscriptions.create("SearchChannel", {
    connected() { // Called when the subscription is ready for use on the server
    },

    disconnected() { // Called when the subscription has been terminated by the server
    },

    received(data) { // Called when there's incoming data on the WebSocket for this channel
        console.log("Received search query:", data.query);
        // You might want to call a function to handle the search data here
        handleSearchData(data.query);
    }
});

document.getElementById('search-input').addEventListener('input', function () {
  console.log("Search input changed:", this.value);
    const query = this.value.trim();

    clearTimeout(bufferTimer);

    if (query.length >= MINIMUM_QUERY_LENGTH) {
      bufferTimer = setTimeout(() => {
        searchChannel.send({command: 'search', query: query});
        }, BUFFERING_DELAY_MS);
    }
});

// Add a function to handle the search data
function handleSearchData(query) { // Implement your logic to handle the search data here
    console.log("Handling search data:", query);
    // For example, update the UI with search results
}

File: app/channels/search_channel.rb
class SearchChannel < ApplicationCable::Channel

  def subscribed
    stream_from "search_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    # Handle incoming messages here
    case data['command']
    when 'search'
      handle_search(data['query'])
    else
      # Handle other commands if needed
    end
  end

  private

  def handle_search(query)
    Rails.logger.info("Received search query: #{query}")
    @most_recent_query = query
    buffer_timer
  end

  def buffer_timer
    @buffer_timer&.cancel
    @buffer_timer = Concurrent::ScheduledTask.execute(1.5) do
      process_search
    end
  end

   def process_search
    final_query = @most_recent_query

    Rails.logger.info("Processing search query: #{final_query}")
  end
end
